src/vos/vos_csum.c

```diff
@@ -0,0 +1,398 @@
+	d_sg_list_t		*cra_sgl;	/* write sgl */
+	struct evt_entry_in	*cra_ent_in;    /* coalesced entry */
+	struct csum_recalc	*cra_recalcs;   /* recalc info */
+	void			*cra_buf;	/* read buffer */
+	daos_size_t		 cra_seg_size;  /* size of coalesced entry */
+	unsigned int		 cra_seg_cnt;   /* # of read segments */
+	unsigned int		 cra_buf_len;	/* length of read buffer */
+	int			 cra_rc;	/* return code */
+	ABT_eventual		 csum_eventual;
+};
+
+/* construct sgl to send to csummer for verify of read data */
+static unsigned int
+csum_agg_set_sgl(d_sg_list_t *sgl, struct bio_sglist *bsgl,
+		 struct csum_recalc *recalcs, uint8_t *buf,
+		 unsigned int buf_len, int add_start, daos_size_t seg_size,
+		 unsigned int idx, unsigned int add_offset,
+		 unsigned int *buf_idx, unsigned int *add_idx)
+{
+	unsigned int sgl_idx = 0;
+
+	sgl->sg_nr = 1;
+	if (recalcs[idx].cr_prefix_len) {
+		sgl->sg_iovs[sgl_idx].iov_buf = &buf[*add_idx+seg_size];
+		D_ASSERT(recalcs[idx].cr_prefix_len ==
+			 bsgl->bs_iovs[add_start + add_offset].bi_data_len);
+		sgl->sg_iovs[sgl_idx].iov_buf_len =
+			bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		sgl->sg_iovs[sgl_idx++].iov_len =
+			bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		*add_idx += bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		add_offset++;
+		sgl->sg_nr++;
+	}
+
+	sgl->sg_iovs[sgl_idx].iov_buf = &buf[*buf_idx];
+	sgl->sg_iovs[sgl_idx].iov_buf_len = bsgl->bs_iovs[idx].bi_data_len;
+	sgl->sg_iovs[sgl_idx++].iov_len = bsgl->bs_iovs[idx].bi_data_len;
+	*buf_idx += bsgl->bs_iovs[idx].bi_data_len;
+
+	if (recalcs[idx].cr_suffix_len) {
+		sgl->sg_iovs[sgl_idx].iov_buf = &buf[*add_idx + seg_size];
+		D_ASSERT(recalcs[idx].cr_suffix_len ==
+			 bsgl->bs_iovs[add_start + add_offset].bi_data_len);
+		sgl->sg_iovs[sgl_idx].iov_buf_len =
+			bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		sgl->sg_iovs[sgl_idx].iov_len =
+			bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		*add_idx += bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		add_offset++;
+		sgl->sg_nr++;
+	}
+	return add_offset;
+}
+
+static unsigned int
+calc_csum_params(struct dcs_csum_info *csum_info, struct csum_recalc *recalc,
+		 unsigned int prefix_len, unsigned int suffix_len,
+		 unsigned int rec_size)
+{
+
+	unsigned int cs_cnt, low_idx = recalc->cr_log_ext.ex_lo -
+							prefix_len / rec_size;
+	unsigned int high_idx = recalc->cr_log_ext.ex_hi +
+							suffix_len / rec_size;
+
+	assert(prefix_len % rec_size == 0);
+	cs_cnt = csum_chunk_count(recalc->cr_phy_ent->pe_csum_info.cs_chunksize,
+				    low_idx, high_idx, rec_size);
+	csum_info->cs_nr = cs_cnt;
+	D_ASSERT(cs_cnt * csum_info->cs_len  <= csum_info->cs_buf_len);
+	return low_idx;
+}
+
+static bool
+csum_agg_verify(struct csum_recalc *recalc, struct dcs_csum_info *new_csum,
+		unsigned int rec_size, unsigned int prefix_len)
+{
+	unsigned int j = 0;
+	bool match;
+
+	if (new_csum->cs_nr != recalc->cr_phy_ent->pe_csum_info.cs_nr) {
```

Comment:
Please add more comments to this code since I don't understand it.

Reply:
Code has been reorganized, with the calculation offload moved to container module.  This allows passing the calculation function into vos_aggregate as a callback function.  (ULT creation is not linked to the VOS library.)  Extensive documentation has been added as comments within the code.